File: 백준/Gold/2580. 스도쿠/스도쿠.js
const board = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((arr) => arr.trim().split(" ").map(Number));

const isValid = (num, row, col) => {
  const inRow = board[row].includes(num);
  const inCol = board.some((row) => row[col] === num);
  const inBox = (num, row, col) => {
    const boxRow = Math.floor(row / 3) * 3;
    const boxCol = Math.floor(col / 3) * 3;
    for (let i = boxRow; i < boxRow + 3; i++) {
      for (let j = boxCol; j < boxCol + 3; j++) {
        if (board[i][j] === num) return true;
      }
    }
    return false;
  };

  return !(inRow || inCol || inBox(num, row, col));
};

const findEmptyPos = () => {
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      if (board[i][j] === 0) return [i, j];
    }
  }
  return null;
};

const backtrack = () => {
  const emptyPos = findEmptyPos();
  if (!emptyPos) return true;

  const [row, col] = emptyPos;

  for (let num = 1; num <= 9; num++) {
    if (isValid(num, row, col)) {
      board[row][col] = num;
      if (backtrack()) return true;
      board[row][col] = 0;
    }
  }

  return false;
};

backtrack();
console.log(board.map((arr) => arr.join(" ")).join("\n"));

File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys
input = sys.stdin.readline

board = [list(map(int, input().split())) for _ in range(9)]

# 각 행, 열, 박스에 있는 숫자를 미리 체크
row = [[False] * 10 for _ in range(9)]
col = [[False] * 10 for _ in range(9)]
box = [[False] * 10 for _ in range(9)]

# 초기 보드 상태 저장
for i in range(9):
    for j in range(9):
        if board[i][j]:
            num = board[i][j]
            row[i][num] = True
            col[j][num] = True
            box[(i//3)*3 + j//3][num] = True

def isValid(num, r, c):
    return not (row[r][num] or col[c][num] or box[(r//3)*3 + c//3][num])

def backtrack(idx):
    if idx == 81:
        return True
        
    r, c = idx // 9, idx % 9
    if board[r][c]:
        return backtrack(idx + 1)
        
    for num in range(1, 10):
        if isValid(num, r, c):
            board[r][c] = num
            row[r][num] = col[c][num] = box[(r//3)*3 + c//3][num] = True
            if backtrack(idx + 1):
                return True
            board[r][c] = 0
            row[r][num] = col[c][num] = box[(r//3)*3 + c//3][num] = False
            
    return False

backtrack(0)
print('\n'.join(' '.join(map(str, row)) for row in board))